### PR TITLE
Improving gas estimates for l2s

### PR DESCRIPTION
**Pull request by snreynolds**
Noteable changes:
- Moved `buildTrade` and `buildSwapMethodParameters` to a separate library so we can access them in the gasModel
- On optimism the gasEstimate now considers the l1security fee
- Added a GasDataProvider that reads from the gas oracle contract to get gas constants for the l1 security fee calculations
- Surfaced `initTicksCrossed`, `gasUseL1`, and `gasCostL1` to the logger so they’re printed at each run of the router (this was to help with manual gas estimation testing but can definitely be removed)
- Moved gasCost constants to be readable by chainId and updated some Arbitrum estimates
- Small nit change: added some logic so that if no protocols are specified in the CLI it defaults to V3 for chains that do not support v2

### CODE REVIEW

File: src/routers/alpha-router/entities/route-with-valid-quote.ts [hunk 1]
```diff
@@ -173,6 +176,9 @@ export class V3RouteWithValidQuote implements IV3RouteWithValidQuote {
   public tradeType: TradeType;
   public poolAddresses: string[];
   public tokenPath: Token[];
+  public initTicksCrossed: BigNumber;
+  public gasUseL1: BigNumber;
+  public gasCostL1: BigNumber;
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/routers/alpha-router/entities/route-with-valid-quote.ts:181]: Yeah I don't actually need any of these values. Was passing them up through the route so I could print them out only at the end for manual testing. I'll just remove

File: src/routers/alpha-router/gas-models/v3/v3-heuristic-gas-model.ts [hunk 2]
```diff
@@ -209,32 +242,61 @@ export class V3HeuristicGasModelFactory extends IV3GasModelFactory {
...
+    );
     const uninitializedTickGasUse = COST_PER_UNINIT_TICK.mul(0);
 
-    const gasUse = BASE_SWAP_COST.add(hopsGasUse)
+    let l1GasUsed = BigNumber.from(0);
+    let l1Fee = BigNumber.from(0);
+    if (chainId == ChainId.OPTIMISM || chainId == ChainId.OPTIMISTIC_KOVAN) {
+      // account for the L1 security fee
+      if (!swapConfig) {
+        log.info('Skipping l1 security fee calculation');
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/routers/alpha-router/gas-models/v3/v3-heuristic-gas-model.ts:265]: ah interesting, that sounds good !

File: cli/base-command.ts [hunk 3]
```diff
@@ -318,5 +320,14 @@ export abstract class BaseCommand extends Command {
...
       gasPriceWei: gasPriceWei.toString(),
     });
+    const hops = routeAmounts[0]?.poolAddresses.length;
+    this.logger.info(`Total hops ${hops}`);
+
+    if (routeAmounts[0]?.protocol == Protocol.V3) {
+      const v3Route = routeAmounts[0] as V3RouteWithValidQuote;
+      this.logger.info(
+        `Total initialized ticks crossed ${v3Route.initializedTicksCrossedList.length}`
+      );
```
Comment (willpote) [inline cli/base-command.ts:330]: Each element in the initializedTIcksCrossedList is the number of ticks crossed for each pool so you need to sum them

File: cli/base-command.ts [hunk 4]
```diff
@@ -318,5 +320,14 @@ export abstract class BaseCommand extends Command {
       estimatedGasUsed: estimatedGasUsed.toString(),
       gasPriceWei: gasPriceWei.toString(),
     });
+    const hops = routeAmounts[0]?.poolAddresses.length;
```
Comment (willpote) [inline cli/base-command.ts:323]: Have to sum across all the routeAmounts

File: src/providers/multicall-uniswap-provider.ts [hunk 5]
```diff
@@ -228,4 +229,101 @@ export class UniswapMulticallProvider extends IMulticallProvider<UniswapMultical
...
+      providerConfig,
+    } = params;
+
+    const gasLimitPerCall =
+      additionalConfig?.gasLimitPerCallOverride ?? this.gasLimitPerCall;
+    const blockNumberOverride = providerConfig?.blockNumber ?? undefined;
+
+    const calls = _.map(functionNames, (functionName, i) => {
+      const fragment = contractInterface.getFunction(functionName);
+      let callData;
```
Comment (willpote) [inline src/providers/multicall-uniswap-provider.ts:261]: const

File: src/providers/multicall-uniswap-provider.ts [hunk 6]
```diff
@@ -228,4 +229,101 @@ export class UniswapMulticallProvider extends IMulticallProvider<UniswapMultical
...
+      const param = functionParams ? functionParams[i] : [];
+      callData = contractInterface.encodeFunctionData(fragment, param);
+      log.debug(`fragment ${fragment}`);
```
Comment (willpote) [inline src/providers/multicall-uniswap-provider.ts:264]: Can remove

**Review by willpote — APPROVED** [review]
Nice! Just nits at this point!

File: cli/base-command.ts (hunk 3 above)
Comment (snreynolds) [reply cli/base-command.ts:330]: Right! Actually these can just be deleted - I just needed these data points when I was manual testing with arbirtrum and optimism.

If you think these are good to print out though in general, I can change it back to summing over all the routes.

**Pull request merged**